odl/operator/operator.py

```diff
@@ -95,57 +80,279 @@ def _default_apply(self, x, out, *args, **kwargs):
     -------
     `None`
     """
-    out.assign(self._call(x, *args, **kwargs))
+    out.assign(op._call_out_of_place(x, **kwargs))
 
 
-class _OperatorMeta(type):
+def _signature_from_spec(func):
+    """Return the signature of a python function as a string.
 
-    """Metaclass used by `Operator` to ensure correct methods.
+    Parameters
+    ----------
+    func : `function`
+        Function whose signature to compile
 
-    If either ``_apply()`` or ``_call()`` does not exist in the class
-    to be created, this metaclass attempts to add a default
-    implmentation.
-    This only works if the `Operator.range` is a
-    `LinearSpace`.
+    Returns
+    -------
+    sig : `str`
+        Signature of the function
     """
+    import sys
+    import inspect
 
-    def __new__(mcs, name, bases, attrs):
-        """Create a new instance."""
-        if '_call' in attrs and '_apply' in attrs:
-            pass
-        elif '_call' in attrs:
-            attrs['_apply'] = _default_apply
-        elif '_apply' in attrs:
-            attrs['_call'] = _default_call
+    # pylint: disable=deprecated-method,redefined-variable-type,no-member
+    py3 = (sys.version_info.major > 2)
+    if py3:
+        spec = inspect.getfullargspec(func)
+    else:
+        spec = inspect.getargspec(func)
+
+    posargs = spec.args
+    defaults = spec.defaults
+    varargs = spec.varargs
+    kwargs = spec.varkw if py3 else spec.keywords
+    deflen = 0 if defaults is None else len(defaults)
+    nodeflen = 0 if posargs is None else len(posargs) - deflen
+    if nodeflen > 0:
+        argstr = ', '.join(posargs[:nodeflen])
+    else:
+        argstr = ''
+
+    if defaults:
+        if argstr:
```

Comment:
this `if argstring:` idiom seems wierd. How about instead:

```python
args = ['{}'.format(arg for arg in posargs[:nodeflen]])
# no need for "if defaults", nodeflen is len(posargs) if defaults is [] so we add empty list
args += ['{}={}'.format(arg, dval)
         for arg, dval in zip(posargs[nodeflen:], defaults)]
# for same reason ignore "if varargs"
args += ['*{}'.format(varargs)]
if py3:
    kw_only = spec.kwonlyargs
    kw_only_defaults = spec.kwonlydefaults
    if kw_only and not varargs:
        args += ['*']
    args += ['{}={}'.format(arg, kw_only_defaults[arg])
             for arg in kw_only]
args += ['**{}'.format(kwargs)]

argstr = ', '.joint(args)
```

I havnt test-run this, but it should work.